kore/src/Kore/Step/SMT/Evaluator.hs

```diff
@@ -173,30 +180,114 @@ goTranslatePredicate
 goTranslatePredicate tools predicate = evalTranslator translator
   where
     translator =
-        give tools $ translatePredicate translateUninterpreted predicate
+        give tools $ translatePredicate translateTerm predicate
 
-translateUninterpreted
+translateTerm
     :: InternalVariable variable
-    => p ~ TermLike variable
     => SMT.MonadSMT m
     => MonadLog m
     => SExpr  -- ^ type name
-    -> p  -- ^ uninterpreted pattern
-    -> Translator m p SExpr
-translateUninterpreted t pat =
-    lookupPattern <|> freeVariable
+    -> TranslateItem variable  -- ^ uninterpreted pattern
+    -> Translator m variable SExpr
+translateTerm smtType (QuantifiedVariable var) = do
+    n <- Counter.increment
+    let varName = "<" <> Text.pack (show n) <> ">"
+        smtVar = SimpleSMT.const varName
+    State.modify'
+        (Lens.over (field @"quantifiedVars") $ Map.insert var
+            SmtEncoding
+            { smtName = varName
+            , smtType
+            , boundVars = []
+            }
+        )
+    return smtVar
+translateTerm t (UninterpretedTerm (TermLike.ElemVar_ var)) =
+    lookupVariable var <|> declareVariable t var
+translateTerm t (UninterpretedTerm pat) = do
+    VarContext { quantifiedVars, terms } <- State.get
+    let freeVars = getFreeVariables $ TermLike.freeVariables pat
+        boundVarsMap =
+            Map.filterWithKey
+                (\k _ -> ElemVar k `Set.member` freeVars)
+                quantifiedVars
+        boundVars = Map.keys boundVarsMap
+        boundPat = TermLike.mkExistsN boundVars pat
+    lookupUninterpreted boundPat quantifiedVars terms
+        <|> declareUninterpreted boundPat boundVars boundVarsMap
```

Comment:
Why does `declareUninterpreted` need `boundVars` in addition to `boundVarsMap`, given that the former is just the keys of the latter?

Reply:
true, I just had it around and thought to use it. I've changed it as you suggested.